Pagination on tall screens

brianlovin opened the pull request:
<!-- FILL OUT THE BELOW FORM OR YOUR PR WILL BE AUTOMATICALLY CLOSED -->
**Status**
- [ ] WIP
- [ ] Ready for review
- [x] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- hyperion (frontend)

**Release notes for users (delete if codebase-only change)**
- Fixes a bug where viewing lists on tall screens would not trigger a pagination event

Closes #2498 

This one turned out to be quite a bit tricker than I thought; would love input and ideas if there's a better way to approach this. The problem is that on tall screens if a feed doesn't reach the bottom of the window, it would be impossible to paginate a list. This is super frustrating for people working on second displays.

This fix attempts to add checks within views containing feeds that triggers a `fetchMore` event if the height of the items returned is less than the height of the scroll container. This behavior *will* repeat as long as the infinite scroll component is not as tall as either the scroll container or the window.

Some of the problems I bumped into:
- Each view needs its own class applied to the `InfiniteList` component to account for views with different layouts and scroll containers
- We have to be careful not to infinite fetch
- We should never refetch while an existing refetch is in-flight, otherwise it's possible to send the same cursor and return duplicate data

This will definitely need thorough testing; the best way to repro is to just visit `localhost:3000` and zoom the viewport out until a feed loads without taking up the height of the window.

As far as I can test locally, this is working well in:
- inbox
- user view
- community view - threads and members
- channel view
- notifications view
- dm threads list

brianlovin committed:
Eslint

brianlovin committed:
Fix infinite scroll loading behavior on tall screens

mxstbr reviewed (commented):
Shouldn't this logic live in the `InfiniteScroll` component?

brianlovin commented:
Hmm...yeah, probably; I found that each view had slightly different requirements for how to handle this refetch, but I'll see if I can come up with a good abstraction

brianlovin committed:
Add forked infinitescroll component

brianlovin committed:
Add check for tallscreens

brianlovin committed:
Remove unused helper

brianlovin committed:
Implement new infinitescroller into views

brianlovin commented:
@mxstbr was kind of at a loss on how to best move this logic inside the `InfiniteScroll` component since we're working off of a PR of a fork of another library...would be a nightmare to maintain a pull off of that pull. So what I did was just snagged the code from you PR and added it to `src/components/infiniteScroll` - then I just moved my helper function to check for tall screens.

It took me a while to wrangle with some logic of preventing re-fetches with the same cursor - you can see that in `infiniteScroll:componentDidUpdate`

I've tested all the views I can think of and this is working well; the main downside of this PR is that it diverts us completely away from the original infinite scroll component (although that hasn't impacted us in the last year, so not sure why it would matter too much now).

brianlovin committed:
Remove console

brianlovin committed:
Fix flow

brianlovin committed:
Fix flow again

mxstbr commented on src/components/infiniteScroll/index.js:
```diff
@@ -1,35 +1,224 @@
+      const currFlipMoveChildren = React.Children.toArray(curr.children)[0]
+        .props.children;
+      const prevFlipMoveChildren = React.Children.toArray(prevProps.children)[0]
+        .props.children;
```
Would you mind adding a comment about wtf this code does? Why do you need to manually manipulate children?

brianlovin replied:
Yeah I'll add comments here; basically I need to avoid a refetch is the same number of children exist at this step; the problem is in the dashboard the `InfiniteScroll` has only one child: the `FlipMove` component. So this code is checking for the number of children that the flipmove component receives, which maps to the number of threads rendered. Overall I know this isn't perfect, but I was really struggling to find any way to prevent the component from calling two `fetchMore`s in a row with the same cursor ID, and this child length equality check does the job.

mxstbr commented on src/components/infiniteScroll/index.js:
```diff
@@ -1,35 +1,224 @@
+      // $FlowFixMe
+      props.ref = node => {
+        this.scrollComponent = node;
+      };
```
And what are we doing here? Why are we attaching something to the props? This should be a callback and it should definitely not live in the render method:

```JS
this.props.innerRef((node) => {});
```

brianlovin replied:
This is an artifact from the original library we are using; happy to take it out, but I'm not sure why the original author had it there and what might break without it.

brianlovin replied:
I have no clue how to approach this to be honest man :/

mxstbr reviewed (commented):
I'm not sure I understand the code that's in there. Mind adding some explanations to it?

brianlovin committed:
Merge branch 'alpha' of github.com:withspectrum/spectrum into pagination-on-tall-screens

brianlovin committed:
Add comments to confusing code

brianlovin commented:
I've added some comments, but not sure how to address the ref issue; might be cool to pair on this as I'm stuck on where and when to add this.

mxstbr reviewed (approved):
No worries about the ref thing if it's an upstream thing—we should look into rebuilding this thing from scratch eventually anyway, it's not very nice.

If it works, let's ship it! /cc @stubailo for the Apollo issue with the `fetchMore`, he might have an idea why that fetches multiple times.

brianlovin commented:
Agreed, sounds good!
